File: mini-nn-python/mini_nn/tensor.py
"""Tensor operations for neural networks."""

import numpy as np
from typing import Tuple, Optional, Callable


class Tensor:
    """A simple 2D tensor wrapper around numpy arrays."""

    def __init__(self, data: np.ndarray):
        """Initialize tensor from numpy array."""
        if data.ndim == 1:
            data = data.reshape(-1, 1)
        self.data = data.astype(np.float64)

    @property
    def shape(self) -> Tuple[int, int]:
        """Return (rows, cols) shape."""
        return self.data.shape

    @property
    def rows(self) -> int:
        """Return number of rows."""
        return self.data.shape[0]

    @property
    def cols(self) -> int:
        """Return number of columns."""
        return self.data.shape[1]

    # Static factory methods
    @staticmethod
    def zeros(rows: int, cols: int) -> "Tensor":
        """Create a tensor filled with zeros."""
        return Tensor(np.zeros((rows, cols)))

    @staticmethod
    def ones(rows: int, cols: int) -> "Tensor":
        """Create a tensor filled with ones."""
        return Tensor(np.ones((rows, cols)))

    @staticmethod
    def xavier(rows: int, cols: int, rng: Optional[np.random.Generator] = None) -> "Tensor":
        """Create a tensor with Xavier initialization."""
        if rng is None:
            rng = np.random.default_rng()
        scale = np.sqrt(2.0 / (rows + cols))
        return Tensor(rng.normal(0, scale, (rows, cols)))

    @staticmethod
    def he(rows: int, cols: int, rng: Optional[np.random.Generator] = None) -> "Tensor":
        """Create a tensor with He initialization."""
        if rng is None:
            rng = np.random.default_rng()
        scale = np.sqrt(2.0 / rows)
        return Tensor(rng.normal(0, scale, (rows, cols)))

    @staticmethod
    def from_array(arr: np.ndarray) -> "Tensor":
        """Create tensor from numpy array."""
        return Tensor(arr)

    # Element access
    def __getitem__(self, key):
        """Get element or slice."""
        result = self.data[key]
        if isinstance(result, np.ndarray):
            if result.ndim == 1:
                result = result.reshape(1, -1)
            return Tensor(result)
        return result

    def __setitem__(self, key, value):
        """Set element or slice."""
        if isinstance(value, Tensor):
            self.data[key] = value.data
        else:
            self.data[key] = value

    def get(self, i: int, j: int) -> float:
        """Get element at (i, j)."""
        return self.data[i, j]

    def set(self, i: int, j: int, value: float):
        """Set element at (i, j)."""
        self.data[i, j] = value

    def get_row(self, i: int) -> np.ndarray:
        """Get row i as 1D array."""
        return self.data[i].copy()

    # Matrix operations
    def matmul(self, other: "Tensor") -> "Tensor":
        """Matrix multiplication."""
        return Tensor(self.data @ other.data)

    def transpose(self) -> "Tensor":
        """Transpose the tensor."""
        return Tensor(self.data.T)

    @property
    def T(self) -> "Tensor":
        """Transpose property."""
        return self.transpose()

    def add(self, other: "Tensor") -> "Tensor":
        """Element-wise addition with broadcasting."""
        return Tensor(self.data + other.data)

    def sub(self, other: "Tensor") -> "Tensor":
        """Element-wise subtraction."""
        return Tensor(self.data - other.data)

    def mul(self, other: "Tensor") -> "Tensor":
        """Element-wise multiplication."""
        return Tensor(self.data * other.data)

    def div(self, other: "Tensor") -> "Tensor":
        """Element-wise division."""
        return Tensor(self.data / other.data)

    def scale(self, scalar: float) -> "Tensor":
        """Scale by scalar."""
        return Tensor(self.data * scalar)

    def apply(self, func: Callable[[np.ndarray], np.ndarray]) -> "Tensor":
        """Apply function element-wise."""
        return Tensor(func(self.data))

    def sum(self) -> float:
        """Sum all elements."""
        return float(np.sum(self.data))

    def sum_axis(self, axis: int) -> "Tensor":
        """Sum along axis."""
        return Tensor(np.sum(self.data, axis=axis, keepdims=True))

    def mean(self) -> float:
        """Mean of all elements."""
        return float(np.mean(self.data))

    def slice_rows(self, start: int, end: int) -> "Tensor":
        """Slice rows from start to end (exclusive)."""
        return Tensor(self.data[start:end].copy())

    def copy(self) -> "Tensor":
        """Create a copy of this tensor."""
        return Tensor(self.data.copy())

    # In-place operations
    def add_inplace(self, other: "Tensor"):
        """In-place addition."""
        self.data += other.data

    def scale_inplace(self, scalar: float):
        """In-place scaling."""
        self.data *= scalar

    def fill(self, value: float):
        """Fill with value."""
        self.data.fill(value)

    # Numpy interop
    def numpy(self) -> np.ndarray:
        """Return underlying numpy array."""
        return self.data

    def __repr__(self) -> str:
        return f"Tensor({self.rows}, {self.cols})\n{self.data}"

    def __add__(self, other: "Tensor") -> "Tensor":
        return self.add(other)

    def __sub__(self, other: "Tensor") -> "Tensor":
        return self.sub(other)

    def __mul__(self, other) -> "Tensor":
        if isinstance(other, Tensor):
            return self.mul(other)
        return self.scale(other)

    def __rmul__(self, other) -> "Tensor":
        return self.scale(other)

    def __matmul__(self, other: "Tensor") -> "Tensor":
        return self.matmul(other)

    def __neg__(self) -> "Tensor":
        return self.scale(-1)

File: mini-nn-java/src/main/java/com/mininn/Tensor.java
package com.mininn;

import java.util.Random;
import java.util.function.DoubleUnaryOperator;

/**
 * A simple 2D tensor (matrix) implementation for neural network operations.
 */
public class Tensor {
    private final double[][] data;
    private final int rows;
    private final int cols;

    public Tensor(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.data = new double[rows][cols];
    }

    public Tensor(double[][] data) {
        this.rows = data.length;
        this.cols = data.length > 0 ? data[0].length : 0;
        this.data = new double[rows][cols];
        for (int i = 0; i < rows; i++) {
            System.arraycopy(data[i], 0, this.data[i], 0, cols);
        }
    }

    // Static factory methods
    public static Tensor zeros(int rows, int cols) {
        return new Tensor(rows, cols);
    }

    public static Tensor ones(int rows, int cols) {
        Tensor t = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                t.data[i][j] = 1.0;
            }
        }
        return t;
    }

    public static Tensor xavier(int rows, int cols, Random rng) {
        Tensor t = new Tensor(rows, cols);
        double scale = Math.sqrt(2.0 / (rows + cols));
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                t.data[i][j] = rng.nextGaussian() * scale;
            }
        }
        return t;
    }

    public static Tensor he(int rows, int cols, Random rng) {
        Tensor t = new Tensor(rows, cols);
        double scale = Math.sqrt(2.0 / rows);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                t.data[i][j] = rng.nextGaussian() * scale;
            }
        }
        return t;
    }

    public static Tensor fromArray(double[] arr, int rows, int cols) {
        Tensor t = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                t.data[i][j] = arr[i * cols + j];
            }
        }
        return t;
    }

    // Getters
    public int getRows() { return rows; }
    public int getCols() { return cols; }
    public int[] getShape() { return new int[]{rows, cols}; }

    public double get(int i, int j) {
        return data[i][j];
    }

    public void set(int i, int j, double value) {
        data[i][j] = value;
    }

    public double[] getRow(int i) {
        double[] row = new double[cols];
        System.arraycopy(data[i], 0, row, 0, cols);
        return row;
    }

    public void setRow(int i, double[] row) {
        System.arraycopy(row, 0, data[i], 0, Math.min(cols, row.length));
    }

    // Matrix operations
    public Tensor matmul(Tensor other) {
        if (this.cols != other.rows) {
            throw new IllegalArgumentException(
                String.format("Matrix dimensions don't match for multiplication: (%d,%d) x (%d,%d)",
                    rows, cols, other.rows, other.cols));
        }
        Tensor result = new Tensor(this.rows, other.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < other.cols; j++) {
                double sum = 0;
                for (int k = 0; k < this.cols; k++) {
                    sum += this.data[i][k] * other.data[k][j];
                }
                result.data[i][j] = sum;
            }
        }
        return result;
    }

    public Tensor transpose() {
        Tensor result = new Tensor(cols, rows);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result.data[j][i] = data[i][j];
            }
        }
        return result;
    }

    public Tensor add(Tensor other) {
        if (other.rows == 1 && other.cols == this.cols) {
            // Broadcasting: add row vector to each row
            Tensor result = new Tensor(rows, cols);
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    result.data[i][j] = data[i][j] + other.data[0][j];
                }
            }
            return result;
        }
        if (this.rows != other.rows || this.cols != other.cols) {
            throw new IllegalArgumentException("Tensor dimensions must match for addition");
        }
        Tensor result = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result.data[i][j] = data[i][j] + other.data[i][j];
            }
        }
        return result;
    }

    public Tensor sub(Tensor other) {
        if (this.rows != other.rows || this.cols != other.cols) {
            throw new IllegalArgumentException("Tensor dimensions must match for subtraction");
        }
        Tensor result = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result.data[i][j] = data[i][j] - other.data[i][j];
            }
        }
        return result;
    }

    public Tensor mul(Tensor other) {
        if (this.rows != other.rows || this.cols != other.cols) {
            throw new IllegalArgumentException("Tensor dimensions must match for element-wise multiplication");
        }
        Tensor result = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result.data[i][j] = data[i][j] * other.data[i][j];
            }
        }
        return result;
    }

    public Tensor scale(double scalar) {
        Tensor result = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result.data[i][j] = data[i][j] * scalar;
            }
        }
        return result;
    }

    public Tensor apply(DoubleUnaryOperator op) {
        Tensor result = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result.data[i][j] = op.applyAsDouble(data[i][j]);
            }
        }
        return result;
    }

    public double sum() {
        double total = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                total += data[i][j];
            }
        }
        return total;
    }

    public Tensor sumAxis(int axis) {
        if (axis == 0) {
            // Sum along rows -> result is 1 x cols
            Tensor result = new Tensor(1, cols);
            for (int j = 0; j < cols; j++) {
                double sum = 0;
                for (int i = 0; i < rows; i++) {
                    sum += data[i][j];
                }
                result.data[0][j] = sum;
            }
            return result;
        } else {
            // Sum along cols -> result is rows x 1
            Tensor result = new Tensor(rows, 1);
            for (int i = 0; i < rows; i++) {
                double sum = 0;
                for (int j = 0; j < cols; j++) {
                    sum += data[i][j];
                }
                result.data[i][0] = sum;
            }
            return result;
        }
    }

    public Tensor sliceRows(int start, int end) {
        Tensor result = new Tensor(end - start, cols);
        for (int i = start; i < end; i++) {
            System.arraycopy(data[i], 0, result.data[i - start], 0, cols);
        }
        return result;
    }

    public Tensor copy() {
        Tensor result = new Tensor(rows, cols);
        for (int i = 0; i < rows; i++) {
            System.arraycopy(data[i], 0, result.data[i], 0, cols);
        }
        return result;
    }

    public void addInPlace(Tensor other) {
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                data[i][j] += other.data[i][j];
            }
        }
    }

    public void scaleInPlace(double scalar) {
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                data[i][j] *= scalar;
            }
        }
    }

    public void fill(double value) {
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                data[i][j] = value;
            }
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("Tensor(%d, %d)\n", rows, cols));
        for (int i = 0; i < Math.min(rows, 5); i++) {
            sb.append("[");
            for (int j = 0; j < Math.min(cols, 5); j++) {
                sb.append(String.format("%8.4f", data[i][j]));
                if (j < cols - 1) sb.append(", ");
            }
            if (cols > 5) sb.append("...");
            sb.append("]\n");
        }
        if (rows > 5) sb.append("...\n");
        return sb.toString();
    }
}
